File: mind.cpp
#include <cmath>

// Forward declarations for Path class
class Path;

// Individual class representing individuals
class Individual {
public:
    Path w;
    double x, y, z;
};

// Declaration for Mind class containing member functions from the 'mind' module
class Mind {
public:
    void makeI(Path w, double x, double y, double z, Individual& i);
    void pathI(const Individual& i, Path& w);
    double xposI(const Individual& i);
    double yposI(const Individual& i);
    double zposI(const Individual& i);
    double distI(const Individual& i1, const Individual& i2);
    bool ind_eqI(const Individual& i1, const Individual& i2);
};

// Implementation of the functions in the Mind class

void Mind::makeI(Path w, double x, double y, double z, Individual& i) {
    i.w = w;
    i.x = x;
    i.y = y;
    i.z = z;
}

void Mind::pathI(const Individual& i, Path& w) {
    w = i.w;
}

double Mind::xposI(const Individual& i) {
    return i.x;
}

double Mind::yposI(const Individual& i) {
    return i.y;
}

double Mind::zposI(const Individual& i) {
    return i.z;
}

double Mind::distI(const Individual& i1, const Individual& i2) {
    double a = i1.x;
    double b = i1.y;
    double c = i1.z;
    double d = i2.x;
    double e = i2.y;
    double f = i2.z;

    return sqrt(((a - d) * (a - d)) + ((b - e) * (b - e)) + ((c - f) * (c - f)));
}

bool Mind::ind_eqI(const Individual& i1, const Individual& i2) {
    return (xposI(i1) == xposI(i2)) && (yposI(i1) == yposI(i2)) && (zposI(i1) == zposI(i2));
}

File: mind.f90
module mind
    ! Declarations
    use mpath

    public :: makeI, pathI, xposI, yposI, zposI, distI, ind_eqI

    type, public :: ind
        private
        type(path) :: w
        real :: x
        real :: y
        real :: z
    end type ind

contains

    ! Subroutine to create an individual
    subroutine makeI(w, x, y, z, i)
        type(path), intent(in) :: w
        real, intent(in) :: x, y, z
        type(ind), intent(out) :: i

        i%w = w
        i%x = x
        i%y = y
        i%z = z
    end subroutine makeI

    ! Subroutine to obtain the path of an individual
    subroutine pathI(i, w)
        type(ind), intent(in) :: i
        type(path), intent(out) :: w

        w = i%w
    end subroutine pathI

    ! Function to get the x-coordinate of an individual
    function xposI(i) result(x)
        type(ind), intent(in) :: i
        real :: x

        x = i%x
    end function xposI

    ! Function to get the y-coordinate of an individual
    function yposI(i) result(y)
        type(ind), intent(in) :: i
        real :: y

        y = i%y
    end function yposI

    ! Function to get the z-coordinate of an individual
    function zposI(i) result(z)
        type(ind), intent(in) :: i
        real :: z

        z = i%z
    end function zposI

    ! Function to calculate the distance between two individuals
    function distI(i1, i2) result(r)
        type(ind) :: i1, i2
        real :: a, b, c, d, e, f, r

        a = i1%x
        b = i1%y
        c = i1%z
        d = i2%x
        e = i2%y
        f = i2%z

        r = sqrt(((a - d) ** 2) + ((b - e) ** 2) + ((c - f) ** 2))
    end function distI

    ! Function to check equality between two individuals
    function ind_eqI(i1, i2) result(b)
        type(ind), intent(in) :: i1, i2
        logical :: b

        if ((xposI(i1) == xposI(i2)) .and. (yposI(i1) == yposI(i2)) .and. (zposI(i1) == zposI(i2))) then
            b = .true.
        else
            b = .false.
        end if
    end function ind_eqI

end module mind
